[리액트 8기 이재윤] 7주차 과제 1

jaeyooniverse opened the pull request:
과거에 했던 다짐을 다시 생각해보며 아직 끝내지 못한 것을 포함하여 더욱 더 열심히 노력하는 한 주가 되겠습니다.
이번 주도 잘 부탁드립니다!

jaeyooniverse committed:
첫 번째 강의 내용 업데이트, 파일 분류

hannut91 commented on src/services/api.js:
```diff
@@ -13,17 +13,31 @@ export async function fetchCategories() {
 }
+
+export async function postLogin({ email, password }) {
+  const url = 'https://eatgo-login-api.ahastudio.com/session';
```
계속 URL이 반복되서 사용하고 있는 것 같네요. 중복을 제거해 보는게 좋을 것 같아요

jaeyooniverse replied:
확실히 제거하는 것이 더 간결하네요! url 중복도 제거하면서 data를 fetch해오는 과정의 중복도 제거해봤습니다.
좋은 리뷰 감사합니다!
done: https://github.com/CodeSoom/react-week7-assignment-1/pull/79/commits/dfa82dc082b9e1ae3f052a073edd2e679c4555c6

hannut91 commented on src/reducer.test.js:
```diff
@@ -125,4 +119,54 @@ describe('reducer', () => {
+          changeLoginField({ name: 'password', value: 'test' })
+        );
+
+        expect(state.loginFields.email).toBe('email');
```
기존과 변하지 않는다는 것을 드러내려면 다음과 같이 할 수도 있을 것 같아요.

```
expect(state.loginFields.email).toBe(initialState.loginFields.email)
```

jaeyooniverse replied:
수정했습니다.
done: https://github.com/CodeSoom/react-week7-assignment-1/pull/79/commits/7a3444e3d5098e64b9fdad8859c5742bc774a18f

jaeyooniverse committed:
두 번째 강의 내용 업데이트

jaeyooniverse committed:
세 번째 강의 내용 업데이트

jaeyooniverse committed:
반복되는 코드, url 리팩토링

이전 코드에서 data를 fetch 해오는 과정과 API 주소의 url이 상당히 많이 중복되어 있었기 때문에
이를 리팩토링하여 코드를 간결하게 만들었습니다.

hannut91 commented on src/LogoutForm.jsx:
```diff
@@ -0,0 +1,9 @@
+export default function LogoutForm({ onClick }) {
```
얘는 실제로 `<form />`이 없어서 `Form`이 아닌 것 같네요. Form이라는 정의가 무엇일까 궁금해지네요.

hannut91 commented on src/LogoutForm.test.jsx:
```diff
@@ -0,0 +1,41 @@
+    expect(handleClick).toBeCalled();
+  });
+
+  context('when button is not clicked', () => {
+    it("doesn't click event handler", () => {
+      const handleClick = jest.fn();
+
+      render(<LogoutForm onClick={handleClick} />);
+
+      expect(handleClick).not.toBeCalled();
+    });
+  });
```
이 테스트는 무엇을 드러내고 싶어서 작성하셨나요?

hannut91 commented on src/containers/Restaurant/RestaurantContainer.jsx:
```diff
@@ -0,0 +1,47 @@
+  return (
+    <>
+      <RestaurantDetail restaurant={restaurant} />
+      {accessToken ? (
+        <ReviewForm
+          fields={reviewFields}
+          onChange={handleChange}
+          onSubmit={handleSubmit}
+        />
+      ) : null}
```
JSX에서 `false` , `true` , `undefined` , `null` 은 아무것도 렌더하지 않습니다. `accessToken`이 기본값이 null인 것 같던데, 다른 값을 가질수도 있는건가요?

See also

- [https://reactjs.org/docs/jsx-in-depth.html#booleans-null-and-undefined-are-ignored](https://reactjs.org/docs/jsx-in-depth.html#booleans-null-and-undefined-are-ignored)

hannut91 commented on src/services/api.js:
```diff
@@ -1,29 +1,59 @@
-export async function fetchRegions() {
-  const url = 'https://eatgo-customer-api.ahastudio.com/regions';
+const BASE_URL = 'https://eatgo-customer-api.ahastudio.com';
```
중복을 잘 제거해 주셨네요. 지금은 하나의 서버에만 연결을 하고 있지만, 만약에 이 값이 어떤 환경에 따라서 다르게 실행되어야 한다면 어떻게 할 수 있을까요?

예를들어서 개발환경에서는 개발용 서버를 바라보고, 실제환경에서는 실제서버를 바라보도록 하려면 어떻게 할 수 있을지 고민해 보세요

jaeyooniverse committed:
로그인 에러 해결을 위한 postLogin 함수 url 수정

로그인 API의 url 주소가 잘못 되어 있어서 로그인이 정상적으로 동작하지 않기 때문에 url을 수정하여
로그인 기능이 정상적으로 동작하도록 수정했습니다.

jaeyooniverse committed:
리뷰를 제대로 불러오는지 확인하기 위한 loadReview 테스트 추가

리뷰를 정상적으로 불러오는지 확인하기 위해 loadReview 액션 테스트 코드를 추가했습니다.
loadReview 액션은 fetchRestaurant로 식당 정보에 담겨있는 리뷰 데이터를 받아와
리뷰 데이터를 토대로 setReviews 액션을 dispatch합니다.
이를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
리뷰를 불러오는 loadReview, 저장하는 setReviews 액션 추가

fetchRestaurnat으로 식당 정보를 받아와 식당 정보에 존재하는 리뷰들을 setReviews를 통해 저장하는 액션을 추가했습니다.

jaeyooniverse committed:
logout, setReviews 액션 리듀서 테스트 추가

logout 액션은 현재 리덕스 스토어에 담겨있는 accessToken값을 빈 값으로 만들고,
setReviews 액션은 restaurant 객체 속 reviews 배열에 리뷰 정보를 저장합니다.
이를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
setReviews 액션 리듀서 추가

action에 payload값을 restaurant 객체 속 reviews 배열에 저장하는 setReviews 액션을 리듀서에 추가했습니다.

jaeyooniverse committed:
리뷰 내용 추가에 따른 RestaurantContainer.test.jsx 수정

과제의 기능 요구 사항인 리뷰 내용 출력을 위해 RestaurantContainer를 수정했습니다.
스토어에 저장되어있는 restaurant 정보를 받아와 Reviews 컴포넌트에 reviews 값을 prop으로
전달합니다.

jaeyooniverse committed:
리뷰 내용을 제대로 출력하는지 테스트하는 Reviews.test.jsx 추가

Review 컴포넌트는 prop으로 받은 reviews가 존재하면 해당 reviews를 작성한 사람의 이름과 내용이 출력되며
reviews가 빈 값이면 등록된 리뷰가 없어요 문구를 출력합니다. 이를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
리뷰 내용을 출력하는 Reviews.jsx

Reviews 컴포넌트는 prop으로 받은 reviews를 id를 기준으로 내림차순 정렬하여 최신 리뷰가 가장 위에
올라와 리뷰를 출력하는 컴포넌트입니다. 만약, reviews에 아무 것도 존재하지 않는다면 등록된 리뷰가
없어요문구를 출력합니다.

jaeyooniverse committed:
필요없는 파일 제거, postReview mocking API 추가

사용하지 않는 LogoutFormContainer 파일을 제거하고, postReview API를 mocking하는 기능을 추가했습니다.

jaeyooniverse committed:
리뷰 불러오기 위한 오타 수정

Reviews에 전달하는 prop의 이름은 reviews여야 했는데 RestaurantContainer에서 reviws라는 이름으로
전달하고 있어서 리뷰가 렌더되지 않았기 때문에 이를 수정하여 리뷰를 정상적으로 불러옵니다.

jaeyooniverse committed:
localStorage에 토큰을 저장하고 불러오는지 확인하기 위한 테스트 추가

saveItem 함수는 localStorage에 accessToken을 저장하는 함수이고,
loadItem 함수는 localStorage에 존재하는 accessToken을 불러오는 함수입니다.
이를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
api 테스트에 사용하기 위한 accessToken 테스트 데이터 추가

api 테스트에 사용하기 위한 accessToken 테스트 데이터를 fixtures에 추가했습니다.

jaeyooniverse committed:
postLogin, postReview가 정상적으로 동작하는지 확인하기 위한 테스트 추가

postLogin은 email과 password 입력값을 바탕으로 로그인을 진행하여 DB에 존재하는 회원이면
accessToken을 발급하는 함수이며, postReview는 로그인한 사용자만 작성해야 하므로 accessToken과
리뷰를 작성하는 식당의 id, 평점, 리뷰 내용을 받아 실행하면 아무 것도 남지 않게 됩니다.
따라서, 이를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
loadReview와 sendReview 액션 테스트 추가

loadReview는 fetchRestaurant를 통해 받아온 식당 정보 객체 속 reviews 배열을 setReviews에 담아주어 dispatch를
진행하는 액션이며, sendReview는 사용자가 로그인 했을 때만 진행해야 하므로 accessToken과 함께 어떤 식당에
리뷰를 다는지 확인하는 restaurantId, 그리고 입력한 평점과 내용을 바탕으로 postReview를 진행하고 리뷰가 정상적으로
post되면 리뷰 내용을 게시글에서도 확인할 수 있어야 하므로 loadReview 액션을 dispatch합니다.
이를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
강의 내용에서 제거한 테스트 코드 제거

강의 내용에서 미처 제거하지 못한 테스트 코드를 제거했습니다.

jaeyooniverse committed:
requestLogin 액션을 테스트하는 테스트 코드 추가

requestLogin은 email과 password 입력 데이터를 토대로 로그인을 진행하며 DB에 존재하는
회원이면 accessToken을 발급받는 setAccessToken을 dispatch합니다.
이를 확인할 수 있는 테스트를 추가했습니다.

jaeyooniverse committed:
comma-dangle, jsx-quotes, implicit-arrow-linebreak, function-paren-newline 수정

마지막 인자에 콤마가 없으면, 새로운 라인을 추가하여 변경사항에 대해서 출력되고, 새로운 인자를
추가할 때 깜빡하고 콤마를 추가하지 않아서 생기는 에러를 미연에 방지할 수 있습니다. 따라서,
컴포넌트들에 빠트린 콤마를 추가했습니다.
jsx의 prop이나 html 태그들 속 속성들은 쌍따옴표를 사용하며, 그 외에 모든 상황에서는 쌍따옴표를
사용하지 않는 것이 권장사항이기에 수정했습니다.
화살표 함수에서 표현식을 다음 줄에 표현하는 것은 지양해야 하기 때문에 표현식을 한 줄에 표현하도록
수정하였으며 코드를 간결하게 만들기 위해 함수의 괄호가 다음 줄에 넘어가지 않도록 수정했습니다.

jaeyooniverse commented:
<img width="645" alt="image" src="https://user-images.githubusercontent.com/103628685/173976044-67184adf-f8f5-4f3b-ac93-75da8b124825.png">
커버리지 100%, e2e 테스트 통과, lint 에러 없음 모두 해결했는데,
Github CI에선 e2e 테스트가 실패로 나오네요...

ㄴ 오 몇 분 후에 갑자기 또 통과가 됐네요?! 주기적으로 테스트하는건가

jaeyooniverse committed:
기존과 변하지 않는다는 것을 드러내기 위한 테스트 코드 수정

password가 변경될 때 email은 변하지 않는다라는 테스트를 문자열 값으로 판별했는데,
기존 state 객체 속에서 loginFields 객체 안 email 값으로 표현하는게 기존과
변하지 않는다는 것을 드러내기 더 좋은 것 같아서 이를 수정했습니다.

hannut91 commented on src/ReviewForm.jsx:
```diff
@@ -0,0 +1,26 @@
+import TextField from './TextField';
+
+export default function ReviewForm({ fields, onChange, onSubmit }) {
```
`form`태그를 사용하는 것은 어떨까요?

hannut91 commented on src/ReviewForm.test.jsx:
```diff
@@ -0,0 +1,63 @@
+
+  beforeEach(() => {
+    handleChange.mockClear();
+    handleSubmit.mockClear();
```
저번과제에서는 설정에서 초기화되도록 하셨던 것 같은데, 여기서는 수동으로 하셨군요. 수동으로 하신다면, `jest.clearAllMocks()`를 사용하셔도 될 것 같아요

jaeyooniverse replied:
수정했습니다.
done: https://github.com/CodeSoom/react-week7-assignment-1/pull/79/commits/e1ed1f674f65e5c50c94d9e23f6e03f5575cb549

hannut91 commented on src/Reviews.jsx:
```diff
@@ -0,0 +1,19 @@
+
+  return (
+    <ul>
+      {sortedReviews.map((review) => (
+        <li key={review.id}>
+          <div>{review.name}</div>
+          <div>{review.score}</div>
+          <div>{review.description}</div>
+        </li>
+      ))}
```
여기도 구조 분해 할당으로 간편하게 할 수 잇을 것 같아요

jaeyooniverse replied:
수정했습니다.
done: https://github.com/CodeSoom/react-week7-assignment-1/pull/79/commits/1a32de4e864b4dc8c35be0bddde0782635c1b024

hannut91 commented on src/pages/Login/LoginForm.jsx:
```diff
@@ -0,0 +1,38 @@
+export default function LoginForm({ fields, onChange, onSubmit }) {
```
여기도 `<form />`태그를 사용하는게 좋을 것 같네요

jaeyooniverse committed:
구조 분해 할당을 활용하여 코드 리팩토링

기존 인자로 받았던 리뷰 객체를 구조 분해 할당을 활용하여 코드 속 반복되는 부분을 제거했습니다.

jaeyooniverse committed:
리뷰를 등록하면 input에 입력되어 있는 값 제거 기능 추가

리뷰를 등록하면 input에 입력되어 있는 값을 changeReviewField 액션을 통해 제거하여
중복 리뷰를 받지 않도록 하는 기능을 추가했습니다.

jaeyooniverse committed:
리뷰 등록 시, input 입력값 제거 기능 추가에 따른 테스트 수정

리뷰 남기기 버튼을 누르면 기존에는 dispatch를 2회 진행하기 때문에 테스트 기댓값이 2였으나,
input 입력창 제거 기능을 추가하면서 changeReviewField 액션을 2회 더 진행하기 때문에
테스트 기대값을 4로 수정했습니다.

jaeyooniverse committed:
jest 설정 변경에 따른 mockClear() 제거

jest.config.js에 clearMocks값을 true로 변경하여 mockClear를 각 테스트마다 자동으로 진행해주기 때문에
테스트 파일에 존재하는 mockClear를 모두 제거했습니다.

jaeyooniverse committed:
JSX에서 null은 렌더하지 않으므로 false를 반환하도록 수정

hannut91 commented on src/containers/Restaurant/RestaurantContainer.jsx:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  function handleSubmit() {
+    dispatch(sendReview({ restaurantId }));
+
+    dispatch(changeReviewField({ name: 'score', value: '' }));
+    dispatch(changeReviewField({ name: 'description', value: '' }));
```
여기서는 `sendReview`만 하고, 초기화 액션은 `sendReview`액션 안에서 해도 좋을 것 같아요. 지금은, 이 컴포넌트에서 리뷰를 전송하면 초기화를 한다는 정보가 노출되고 있는 것 같아요. 이 컴포넌트는 액션만 발행할 뿐이지 이러한 정보는 모르는게 좋습니다.

jaeyooniverse replied:
아.. 기능만 생각하다 정작 더 중요한 관심사의 분리를 놓쳤네요!
좋은 리뷰 감사합니다!

hannut91 reviewed (commented):
에러처리, 환경 변수로 분리하기만 되면 완벽할 것 같아요 ㅎㅎ

jaeyooniverse commented:
윤석님 리뷰대로 개발 서버와 배포 서버를 분리하기 위해 .env를 도입했는데 생각해보니 Github CI/CD는 환경 변수를 모르니까 테스트를 통과하지 않게 되겠네요. 이럴땐 어떻게 해결해야 하나요?

jaeyooniverse commented:
끝내 해결하지 못한 것이 에러 핸들링이였는데 이번 주말에 다른 분들의 에러 핸들링 코드를 참고하면서 공부해본 후 8주차에 적용해보겠습니다. 한 주 동안 수고 많으셨습니다! 윤석님!

hannut91 commented:
그러네요 ㅎㅎ 환경 변수가 없어서 GitHub Actions에서는 실패하겠네요. 환경 변수를 추가해야겠어요.

hannut91 commented:
`.github/workflows/ci.yml`파일에 환경 변수를 추가해주고 (CUSTOMER_API_URL, AUTH_API_URL은 제가 지은 이름입니다.)

```
name: CI

on: [push, pull_request]

env:
  CUSTOMER_API_URL: https://eatgo-customer-api.ahastudio.com
  AUTH_API_URL: https://eatgo-login-api.ahastudio.com

jobs:
// ... 생략
```

`webpack.config.js`파일에서 다음과 같이 설정값을 주면

```js
  plugins: [
    new Dotenv({
      systemvars: true,
    }),
  ],
```

동작하네요. 여기서 `systemvars: true`은 `.env`파일에서 읽을 뿐만 아니라, 환경변수에서 읽을 수 있도록 하는 옵션입니다.

See also
  - https://github.com/mrsteele/dotenv-webpack#properties

hannut91 commented on .github/workflows/ci.yml:
```diff
@@ -2,6 +2,10 @@ name: CI
 
+env:
+  CUSTOMER_API_URL: https://eatgo-customer-api.ahastudio.com
+  AUTH_API_URL: https://eatgo-login-api.ahastudio.com
```
제가 환경변수 이름을 재윤님과 다르게 사용해서, 여기 환경 변수 이름을 재윤님이 사용하고 있는 변수 이름과 동일하게 해주어야 합니다

jaeyooniverse committed:
오류 수정

지금까지 dotenv 설정을 잘못 하고 있어서 그와 관련된 코드를 전면 수정했습니다.
